### ISSUE TITLE
Incorrect evaluation of type compatibility involving string literal types as generic type parameters

### ISSUE THREAD

**Issue by JohannesRudolph** [Microsoft/TypeScript#13385]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)
2.1.4

**Code**

```ts
interface HalObjectLinks {
  self: HalLink;
  [rel: string]: HalLink;
}

interface HalLink {
  href: string;
}

interface HalObject<R> {
  '_links': HalObjectLinks;
  // other properties of the object
  [key: string]: any;
}

type CustomerRels = 'addresses';

interface Customer extends HalObject<CustomerRels> {
  name: string;
}


class CustomerModel implements Customer {
  public name: string;

  public '_links' = {
    self: { href: 'self' }
  };
}

function findHref<R extends string>(hal: HalObject<R>, rel: R): string {
  let link = hal._links[<string>rel];
  if (!link) {
    throw new Error('HAL object did not have link: ' + rel);
  }

  return link.href;
}

let x = new CustomerModel();
findHref(x, 'test');
```

**Expected behavior:**
TSC should flag the `findHref()` call for incorrect types. 

**Actual behavior:**
The compiler types the call as: 
```ts
function findHref<"test">(hal: HalObject<"test">, rel: "test"): string
```

However, x is clearly not compatible to `HalObject<"test">`, it's rather compatible to `HalObject<"addresses">`.

The compiler flags no type error.

**Comment by HerringtonDarkholme** [comment]
Please first read the FAQ. https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-type-inference-work-on-this-interface-interface-foot---

TypeScript has structural typing which means 

```ts

interface HalObject<R> {
  '_links': HalObjectLinks;
  // other properties of the object
  [key: string]: any;
}
```
is equivalent to having no type parameter at all.

**Comment by JohannesRudolph** [comment]
Thanks for the pointer to the FAQ, I was indeed missing that. I realise this should have better been a question on Stackoverflow for modelling this, but coming from a C# Background structural typing feels unintuitive when TS does otherwise a very good job of making me productive.
